tests/cloudpickle_test.py

```diff
@@ -2352,6 +2371,71 @@ def test_lookup_module_and_qualname_stdlib_typevar():
     assert name == 'AnyStr'
 
 
+def test_lookup_module_and_qualname_external_module_remove():
+    import _cloudpickle_testpkg
+    T = _cloudpickle_testpkg.T
+
+    register_pickle_by_value(_cloudpickle_testpkg)
+    unregister_pickle_by_value(_cloudpickle_testpkg)
+    module_and_name = _lookup_module_and_qualname(T, name=T.__name__)
+    assert module_and_name is not None
+
+
+def test_lookup_module_and_qualname_external_module_remove_name():
+    import _cloudpickle_testpkg
+    T = _cloudpickle_testpkg.T
+
+    register_pickle_by_value("_cloudpickle_testpkg")
+    unregister_pickle_by_value("_cloudpickle_testpkg")
+    module_and_name = _lookup_module_and_qualname(T, name=T.__name__)
+    assert module_and_name is not None
+
+
+def test_lookup_module_and_qualname_external_module():
+    import _cloudpickle_testpkg
+    T = _cloudpickle_testpkg.T
+
+    register_pickle_by_value(_cloudpickle_testpkg)
+    module_and_name = _lookup_module_and_qualname(T, name=T.__name__)
+    unregister_pickle_by_value(_cloudpickle_testpkg)
+    assert module_and_name is None
+
+
+def test_lookup_module_and_qualname_external_module_name():
+    import _cloudpickle_testpkg
+    T = _cloudpickle_testpkg.T
+
+    register_pickle_by_value("_cloudpickle_testpkg")
+    module_and_name = _lookup_module_and_qualname(T, name=T.__name__)
+    unregister_pickle_by_value("_cloudpickle_testpkg")
+    assert module_and_name is None
+
+
+def test_deep_serialization_package_parsing_parents():
+    # We should deep copy children of explicit modules, not parents
+    package = "foo.bar.baz"
+    register_pickle_by_value(package)
+    result = _is_explicitly_serialized_module("foo.bar")
+    unregister_pickle_by_value(package)
+    assert not result
+
+
+def test_deep_serialization_package_parsing_children():
+    package = "foo.bar"
+    register_pickle_by_value(package)
+    result = _is_explicitly_serialized_module("foo.bar.baz")
+    unregister_pickle_by_value(package)
+    assert result
+
+
+def test_deep_serialization_package_parsing_children_no_submodule():
```

Comment:
maybe these tests should include a `try... finally` block to make sure the global variable:
`_PICKLE_BY_VALUE_MODULES` is restored to its empty value at the end of each test? Unless that is redundant because of some test isolation logic done by `pytest` that I am not aware of.

Reply:
Ive added explicit checks to ensure the _PICKLE_BY_VALUE_MODULES is empty after the tests

Reply:
Correct me if I am wrong, but I think the `assert len(_PICKLE_BY_VALUE_MODULES) == 0` statements are not enough to prevent cross-test side effects: if the assertion actually raises, then the global variable `_PICKLE_BY_VALUE_MODULES` will not be restored to `set()` at the end of the test - hence the use of a cleanup logic within a `finally` clause at the end of such tests.

Reply:
Ah, gotcha. I'll add that in now.